### ISSUE TITLE
Feature Request - event for when record update is sent and ACK:ed by the server

### ISSUE THREAD

**Issue by ronag** [deepstreamIO/deepstream.io-client-js#179]
e.g.
```javascript
app.post('items', wrap(async (req, res, next) => {
   const id = cuid()
   const rec = ds.record.getRecord(id)
   rec.set(req.body)
   // Ensure server has received and ack:ed record update.
   await rec.whenFlushed() // Send record version to server and wait for ACK.
   // NOTE: Only return id when we know (with reasonable certainty) that the record exist in deepstream.
   res.send(id)
})
```

**Comment by WolframHempel** [comment]
This collides with a fundamental design principle. Like most performance focused protocols deepstream is based on eventual consistency and NACKs for writes - negative feedback if something went wrong, rather than acknowledgement of everything successful. Therefor there are no acknowledgements for writes, just errors if something went wrong.

**Comment by ronag** [comment]
@WolframHempel: An ACK is not really required per se. But it would be nice to have at least a fence so you can check whether all commands sent to the server at a certain point have been processed or not.

**Comment by WolframHempel** [comment]
I know what you mean and to an extend this is addressed by the mechanism for #236. But in a distributed system this is trickier than it sounds. Has a write been processed when it reached the first server, the cache or every server within the cluster? Is a write complete only once every client within the system is up to date?

Again, I know that this is something that is asked frequently - and for a database with strong consistency, write locks and master/slave replication this is applicable. For a decentralized system with peer-to-peer replication, NACKs and eventual consistency, determining a write is surprisingly ambiguous

**Comment by ronag** [comment]
@WolframHempel: I think you ar slightly misunderstanding. I totally agree with you that having fences for writes in a distributed system  is difficult. What I'm suggesting is simply a fence that the server which a client is connected to has received all sent messages, nothing else.

**Comment by ronag** [comment]
I want to avoid writing code like this:

```javascript
var rec = ds.record.getRecord('foo')
rec.set({ a: 1 })
new Promise((resolve, reject) => {
  ds.on('error', reject);
  rec.on('error', reject);
  setTimeout(() => resolve(), 2000); // Server probably has received message if no error after 2s.
})
```

**Comment by WolframHempel** [comment]
yup, but that would mean sending acks for writes, thus effectively doubling the number of messages for that scenario. I understand what you're trying to do, but it fundamentally conflicts with deepstream's NACK or "everything's ok unless stated otherwise" paradigm.

**Comment by ronag** [comment]
@WolframHempel: I stil disagree. Consider:

```javascript
for (var n = 0; n < 1000; ++n) {
  var rec = ds.record.getRecord('foo')
  rec.set({ a: 1 })
  rec.discard()
}
// send `PING`
// receive `PONG`
```
Fundamentally what I want can be achieved with a simple heartbeat PING/PONG.

**Comment by ronag** [comment]
Note that I don't want a fence in the since that it guarantees read after write consistency, just a weak guarantee that no network error occurred and eventual consistency will be reached.
